### PR TITLE
Fix issue #315: Get the j-th hessian of the constraints

### PR DESCRIPTION
**Pull request by tmigot**
create the functions:
- `jth_hess_coord`
- `jth_hess_coord!`
- `jth_hess`
and the attribute `neval_jhess` in the `Counters`.

test and document:
- `jth_hess_coord`
- `jth_hess_coord!`
- `jth_hess`
- `jth_hprod`
- `ghjvprod`

### CODE REVIEW

File: src/NLPModels.jl [hunk 1]
```diff
@@ -419,20 +420,83 @@ function jac_op!(nlp :: AbstractNLPModel, x :: AbstractVector, rows :: AbstractV
...
   @lencheck nlp.meta.nvar x v
-  hv = Vector{eltype(x)}(undef, nlp.meta.nvar)
-  return jth_hprod!(nlp, x, v, j, hv)
+  @assert 0 ≤ j ≤ nlp.meta.ncon
+  Hv = Vector{eltype(x)}(undef, nlp.meta.nvar)
+  return jth_hprod!(nlp, x, v, j, Hv)
 end
 
+"""
+    Hv = jth_hprod!(nlp, x, v, j, Hv; obj_weight=1.0)
```
Comment (dpo) [inline src/NLPModels.jl:436]: It doesn't seem to make sense to have `obj_weight` in this method.

File: src/NLPModels.jl [hunk 2]
```diff
@@ -419,20 +420,83 @@ function jac_op!(nlp :: AbstractNLPModel, x :: AbstractVector, rows :: AbstractV
...
+
+Evaluate the product of the j-th Hessian of the constraints with the vector `v`
+in place.
+"""
 function jth_hprod! end
 
+"""
+    vals = jth_hess_coord(nlp, x, j; obj_weight=1.0)
```
Comment (dpo) [inline src/NLPModels.jl:444]: Same here.

File: src/NLPModels.jl [hunk 3]
```diff
@@ -419,20 +420,83 @@ function jac_op!(nlp :: AbstractNLPModel, x :: AbstractVector, rows :: AbstractV
...
+"""
+function jth_hess_coord(nlp::AbstractNLPModel, x::AbstractVector, j::Integer; obj_weight = one(eltype(x)))
+ @lencheck nlp.meta.nvar x
+ @assert 0 ≤ j ≤ nlp.meta.ncon
+ vals = Vector{eltype(x)}(undef, nlp.meta.nnzh)
+ return jth_hess_coord!(nlp, x, j, vals; obj_weight=obj_weight)
+end
+
+"""
+    vals = jth_hess_coord!(nlp, x, j, vals; obj_weight=1.0)
```
Comment (dpo) [inline src/NLPModels.jl:458]: And here.

Comment (dpo) [reply src/NLPModels.jl:458]: What should be the size of `vals` here? Is there an assumption that the same overall `nnzh` is used for all constraint Hessians?

File: src/NLPModels.jl [hunk 4]
```diff
@@ -419,20 +420,83 @@ function jac_op!(nlp :: AbstractNLPModel, x :: AbstractVector, rows :: AbstractV
...
+
+Evaluate the j-th Hessian at `x` scaled by `obj_weight` in sparse coordinate 
+format in place.
+Only the lower triangle is returned.
+"""
+function jth_hess_coord! end
+
+"""
+   Hx = jth_hess(nlp, x, j; obj_weight = 1.0)
```
Comment (dpo) [inline src/NLPModels.jl:467]: obj_weight

File: src/NLPModels.jl [hunk 5]
```diff
@@ -419,20 +420,83 @@ function jac_op!(nlp :: AbstractNLPModel, x :: AbstractVector, rows :: AbstractV
...
+
+Evaluate the j-th Hessian at `x` scaled by `obj_weight` as a sparse matrix.
+Only the lower triangle is returned.
+"""
+function jth_hess(nlp::AbstractNLPModel, x::AbstractVector, j::Integer; obj_weight = one(eltype(x)))
```
Comment (dpo) [inline src/NLPModels.jl:472]: obj_weight

File: src/NLPModels.jl [hunk 6]
```diff
@@ -419,20 +420,83 @@ function jac_op!(nlp :: AbstractNLPModel, x :: AbstractVector, rows :: AbstractV
...
+ @lencheck nlp.meta.nvar x
+ rows, cols = hess_structure(nlp)
+ vals = jth_hess_coord(nlp, x, j, obj_weight=obj_weight)
+ sparse(rows, cols, vals, nlp.meta.nvar, nlp.meta.nvar)
+end
+
+"""
+   gHv = ghjvprod(nlp, x, g, v)
+
+Return the vector of vector-matrix-vector product to each hessian matrices of
```
Comment (dpo) [inline src/NLPModels.jl:484]:
```suggestion
```

File: src/NLPModels.jl [hunk 7]
```diff
@@ -419,20 +420,83 @@ function jac_op!(nlp :: AbstractNLPModel, x :: AbstractVector, rows :: AbstractV
...
+the constraint function, i.e. the vector whose j-th component is g' ∇²cᵢ(x) v.
```
Comment (dpo) [inline src/NLPModels.jl:485]:
```suggestion
Return the vector whose j-th component is g' ∇²cᵢ(x) v.
```

File: src/NLPModels.jl [hunk 8]
```diff
@@ -419,20 +420,83 @@ function jac_op!(nlp :: AbstractNLPModel, x :: AbstractVector, rows :: AbstractV
...
   gHv = Vector{eltype(x)}(undef, nlp.meta.ncon)
   return ghjvprod!(nlp, x, g, v, gHv)
 end
 
+"""
+   ghjvprod!(nlp, x, g, v, gHv)
+
+Return the vector of vector-matrix-vector product to each hessian matrices of
+the constraint function in place, i.e. the vector whose 
+j-th component is g' ∇²cᵢ(x) v.
```
Comment (dpo) [inline src/NLPModels.jl:498]: Fix doc similarly.

File: src/autodiff_model.jl [hunk 9]
```diff
@@ -241,3 +241,38 @@ function hprod!(nlp :: ADNLPModel, x :: AbstractVector, y :: AbstractVector, v :
   Hv .= ForwardDiff.derivative(t -> ForwardDiff.gradient(ℓ, x + t * v), 0)
   return Hv
 end
+
+function jth_hess_coord!(nlp :: ADNLPModel, x :: AbstractVector, j :: Integer, vals :: AbstractVector; obj_weight :: Real = one(eltype(x)))
```
Comment (dpo) [inline src/autodiff_model.jl:245]: Similar comments in this file.

File: test/runtests.jl [hunk 10]
```diff
@@ -144,4 +145,4 @@ include("test_feasibility_form_nls.jl")
 include("test_view_subarray.jl")
 test_view_subarrays()
 include("test_memory_of_coord.jl")
-test_memory_of_coord()
\ No newline at end of file
+test_memory_of_coord()
```
Comment (dpo) [inline test/runtests.jl:148]: I think a newline disappeared here.

File: test/consistency.jl [hunk 11]
```diff
@@ -381,6 +381,65 @@ function consistent_functions(nlps; rtol=1.0e-8, exclude=[])
         end
       end
     end
+    
+    if !(jth_hess_coord in exclude)
+      k = m
+      Ls = Vector{Any}(undef, N)
+      for i = 1:N
+        V = jth_hess_coord(nlps[i], x, y, k)
+        I, J = hess_structure(nlps[i])
```
Comment (dpo) [inline test/consistency.jl:390]: It seems we assume all Hessians have the same structure? That's a constraint that should be well documented.

**Review by dpo — CHANGES_REQUESTED** [review]

File: src/NLPModels.jl [hunk 12]
```diff
@@ -419,20 +420,81 @@ function jac_op!(nlp :: AbstractNLPModel, x :: AbstractVector, rows :: AbstractV
...
+
+Evaluate the product of the j-th Hessian of the constraints with the vector `v`
+in place.
+"""
 function jth_hprod! end
 
+"""
+    vals = jth_hess_coord(nlp, x, j)
+
+Evaluate the j-th Hessian at `x` scaled by `obj_weight` in sparse coordinate
```
Comment (dpo) [inline src/NLPModels.jl:446]: Still has some obj_weight here and there

File: src/NLPModels.jl [hunk 13]
```diff
@@ -419,20 +420,81 @@ function jac_op!(nlp :: AbstractNLPModel, x :: AbstractVector, rows :: AbstractV
...
+   Hx = jth_hess(nlp, x, j)
+
+Evaluate the j-th Hessian at `x` scaled by `obj_weight` as a sparse matrix with
+the same sparsity pattern as the Lagrangian Hessian.
+Only the lower triangle is returned.
+"""
+function jth_hess(nlp::AbstractNLPModel, x::AbstractVector, j::Integer) 
+ @lencheck nlp.meta.nvar x
+ @assert 0 ≤ j ≤ nlp.meta.ncon
+ @lencheck nlp.meta.nvar x
```
Comment (dpo) [inline src/NLPModels.jl:476]: Duplicate of line 474

File: src/NLPModels.jl [hunk 14]
```diff
@@ -419,20 +420,81 @@ function jac_op!(nlp :: AbstractNLPModel, x :: AbstractVector, rows :: AbstractV
...
+ rows, cols = hess_structure(nlp)
+ vals = jth_hess_coord(nlp, x, j)
+ sparse(rows, cols, vals, nlp.meta.nvar, nlp.meta.nvar)
+end
+
+"""
+   gHv = ghjvprod(nlp, x, g, v)
+
+Return the vector whose j-th component is g' ∇²cᵢ(x) v.
```
Comment (dpo) [inline src/NLPModels.jl:485]: i-th ?

File: test/runtests.jl [hunk 15]
```diff
@@ -144,4 +145,4 @@ include("test_feasibility_form_nls.jl")
...
```
Comment (dpo) [inline test/runtests.jl:148]: Newline.

**Review by dpo — CHANGES_REQUESTED** [review]

File: test/consistency.jl (hunk 11 above)
Comment (tmigot) [reply test/consistency.jl]: We assume that the structure is the same than the Hessian Lagrangian, and so in coo formal the vectors have length `nnzh`.
This is now document in the functions `jth_hess_coord!` and `jth_hess`.
In the actual implementation of the function for ADNLPModel and SlackModel, we assert it with `@lencheck nlp.meta.nnzh vals`

**Comment by tmigot** [comment]
It remains to implement the new functions for the variants of NLSModels such as `ADNLSModel`, `FeasibilityFormNLS`, `FeasibilityResidual` and `LLSModel`.

**Comment by tmigot** [comment]
The hessian-related functions are not implemented for lls_model, it will follow progresses in the pull request #282. Nothing difficult here as it is 0 all the time.

**Pull request closed without merging**